apis/Google.Longrunning/Google.Longrunning/Operation.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            get
+            {
+                switch (Proto.ResultCase)
+                {
+                    case Operation.ResultOneofCase.Error:
+                        var error = Proto.Error;
+                        // TODO: Construct a new exception type?
+                        throw new RpcException(new Status((StatusCode) error.Code, error.Message));
+                    case Operation.ResultOneofCase.Response:
+                        return Proto.Response.Unpack<T>();
+                    default:
+                        throw new InvalidOperationException();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Polls the operation until it is complete, returning the completed operation.
+        /// </summary>
+        /// <remarks>
+        /// If this object represents a completed operation
+        /// </remarks>
+        public Operation<T> PollUntilCompleted(PollSettings pollSettings = null)
+        {
+            var current = this;
+            while (!current.IsCompleted) // TODO: Use pollSettings
+            {
+                Thread.Sleep(1000); // TODO: Use pollSettings
+                current = Poll();
+            }
+            return current;
+        }
+
+        public async Task<Operation<T>> PollUntilCompletedAsync(PollSettings pollSettings = null)
+        {
+            var current = this;
+            while (!current.IsCompleted) // TODO: Use pollSettings
+            {
+                await Task.Delay(1000).ConfigureAwait(false); // TODO: Use pollSettings
+                current = await PollAsync().ConfigureAwait(false);
+            }
+            return current;
+        }
+
+        /// <summary>
+        /// Returns a new instance reflecting the most recent state of the operation.
+        /// </summary>
+        public Operation<T> Poll(CallSettings callSettings = null)
+        {
+            var operation = Client.GetOperation(Name, callSettings);
+            return new Operation<T>(operation, Client);
+        }
+
+        public async Task<Operation<T>> PollAsync(CallSettings callSettings = null)
+        {
+            var operation = await Client.GetOperationAsync(Name, callSettings).ConfigureAwait(false);
+            return new Operation<T>(operation, Client);
+        }
+
+        public async Task<Operation<T>> PollAsync(CancellationToken cancellationToken)
+        {
+            var operation = await Client.GetOperationAsync(Name, cancellationToken).ConfigureAwait(false);
+            return new Operation<T>(operation, Client);
+        }
+
+        public void Cancel(CallSettings callSettings = null)
+        {
+            Client.CancelOperation(Name, callSettings);
+        }
+
+        public Task CancelAsync(CallSettings callSettings = null) =>
+            Client.CancelOperationAsync(Name, callSettings);
+
+        // Be careful! The cancellation token is for the cancel operation, so you can cancel cancelling...
```

Comment:
I hope someone makes use of this in a UI by having a "Just kidding" button next to a "Cancel" button.